v2 activity tracking spec (#39)
CloudVLab/qwiklabs-content-bundle-spec: lab-bundle-spec.md

Activity tracking is a feature for evaluating a student's performance in a lab by running a script at "checkpoints". These scripts can call APIs relevant to any environment resource to query their current state. For example, the script may inspect and validate the configuration of GCE instances running in `my-project`, to ensure the user is following the instructions properly.

A lab has an Assessment, which in turn contains the Steps (checkpoints).

#### Assessment

attribute          | required | type    | notes
-------------------| -------- | --------| --------------------------------------
passing_percentage | ✓        | integer | The percentage of total points the student must achieve to "pass" the lab.
steps              | ✓        | array   | An array of [Steps](#steps)

```yml
assessment:
  passing_percentage: 75
  steps: ...
```

#### Steps

attribute        | required | type                       | notes
-----------------| -------- | ---------------------------| --------------------------------------
title            | ✓        | locale dictionary          |
maximum_score    | ✓        | integer                    | The maximum number of points this step can award.

> student_messages | ✓        | array of objects           | The keys are how the messages will be referenced in the code, and the values are locale dictionaries.

Reviewer: i think this should be a dictionary of objects instead of an array of objects

Reviewer 2: Agreed

> student_messages | ✓        | array of objects           | The keys are how the messages will be referenced in the code, and the values are locale dictionaries.

Reviewer 2: ```suggestion
student_messages | ✓        | dictionary of locale dictionaries           | The keys are how the messages will be referenced in the code, and the values are locale dictionaries.
```

services         | ✓        | array of resource services | An array of services that will be used in the code block. Each resource type specifies a set of allowed services.
code             | ✓        | string                     | Code to be executed. See [below](#code) for more information.

```yml
assessment:
  passing_percentage: 75
  steps:
    - title:
        locales:
          en: Create a Cloud Storage bucket
          es: Crear un depósito de almacenamiento en la nube

>       maximum_score: 5

Reviewer 3: I know you removed the pass/fail checkbox based on its lack of usage but I'm curious how many steps offer partial points.

Instead of only allowing integer scores and passing percentages, I'd like to propose only allowing pass/fail steps and a passing step percentage (which defaults to all of them in Git Whisperer).

Author: it's not just a question of how many steps offer partial points - this would also change things if a given assessment has steps with different `max`s. it would also change how game points are calculated. i think i'd rather leave as-is for now, unless others feel strongly.

Reviewer 4: Is there a reason the key is `maximum_score`, but in the code example below we pass in `max_score`?

Author: in the code example it used to be a positional param so it didn't matter, but consistency would probably have been nice. it's now a keyword param and therefore does matter, and has been updated accordingly.

      student_messages:
        - success:
            locales:
              en:
                Great job! You created the bucket!
              es:
                ¡Gran trabajo! ¡Creaste el cubo!
        - bucket_missing:
            locales:
              en:
                Oops! No bucket found.
              es:
                ¡Uy! No se ha encontrado el cubo.
        - bucket_misconfigured:
            locales:
              en:
                Hmm. The bucket is there, but it is misconfigured.
              es:
                Hmm. El cubo está allí, pero está mal configurado.

>       services:

Reviewer 3: I'd like creators to specify a per-step field and I don't know what to call it but basically I'd like creators to specify what granularity their step applies to:

* `zone`: Run this in every GCP zone (e.g. checking if a GCE instance was created in that zone)
* `region`: Run this in every GCP region (e.g. checking if a regional GKE cluster was made in that region)
* `multiregion`: Run this once for US, EU, and Asia (e.g. checking for multi-regional GCS buckets or some newer BigQuery data locality stuff)
* `global`: Run this function once (e.g. checking IAM configuration)

Author: interesting. i don't think we want to run any of this stuff in every zone, and even running it in every region has caused us headaches. i would much rather keep everything "global" until we have a really compelling use case and/or rearchitect things to make "run n times" much easier.

        - target_project.StorageV1

>       code: |-

Reviewer 3: I'd like to also include a `type` or `code_type` field that indicates this step specifies its code as a Ruby method with particular arguments. GitWhisperer could infer it during bundling until we support more types but I think that would make it easier to extend this particular part without messing with the rest of the metadata (and revving the spec version) in the future.

Author: I'm open to that if others are strongly in favor of it, but my preference would be to leave it out for now. If/when it is useful, we can add a field for that in a backwards-compatible way (i.e. without changing the spec version) by just assuming `ruby` if not specified.

>         def check(handles, max_score)

Reviewer 2: @joncass @supersam654 

```suggestion
        def check(handles:handles, resources: resources, maximum_score: maximum_score)
```

          storage_handle = handles['target_project.StorageV1']

          # Check for bucket
          found_bucket = ...
          unless found_bucket
            return { score: 0, student_message: 'bucket_missing' }
          end

          # Check bucket configuration
          bucket_configured_correctly = ...
          unless bucket_configured_correctly
            return { score: 2, student_message: 'bucket_misconfigured' }
          end

>           { score: max_score, student_message: 'success' }

Author: ```suggestion
          { score: maximum_score, student_message: 'success' }
```

        end
    - title:
        locales:
          en: Copy a file to the bucket
          es: Copiar un archivo al cubo
      maximum_score: 5
      student_messages:
        - success:
            locales:
              en:
                Great job! You copied the file!
              es:
                ¡Gran trabajo! ¡Copiaste el archivo!
        - file_missing:
            locales:
              en:
                Oops! No file found.
              es:
                ¡Uy! No se ha encontrado el archivo.
        - file_mismatch:
            locales:
              en:
                Hmm. There's a file here, but it doesn't match the source contents.
              es:
                Hmm. Hay un archivo aquí, pero no coincide con el contenido de origen.
      services:
        - source_project.StorageV1
        - target_project.StorageV1
      code: |-

>         def check(handles, max_score)

Reviewer: something that the current activity tracking code runner does is augment these handles with :projects and :users for GCP api and :domain and :users for GSuite api. we need to figure out how we want to provide these for v2 lab activity tracking, as we are no longer providing an API field

>         def check(handles, max_score)

Reviewer 2: @joncass, @supersam654 

```suggestion
        def check(handles:handles, resources: resources, maximum_score: maximum_score)
```

>           target_storage_handle = handles['target_project.StorageV1']

Reviewer: do we want to pass handles in this format, or more like ```handles[:target_project][:StorageV1]```?
also we will probably need to list the possible services for each resource type

Author: i'd be fine with passing handles in either format - what do you think?
and yeah I have it as a "TODO" in the PR description to enumerate the services for each resource type.

Reviewer 3: Is changing the contract of the arguments to this method in scope for a v2 bundle change? Alternatively, could we make a metadata change for all of the other fields and then talk about how we want the code snippet to function?

Author: yes, changing the contract of the arguments is in scope - that's actually the main reason for taking this on now: that multi-resource labs necessarily need a different signature

Reviewer 2: TODO: Agreed. Why not both?

          source_storage_handle = handles['source_project.StorageV1']

          # Check for file
          found_file = ...
          unless found_file
            return { score: 0, student_message: 'file_missing' }
          end

          # Check file contents
          source_contents = ...
          target_contents = ...
          unless source_contents == target_contents
            return { score: 2, student_message: 'file_mismatch' }
          end

>           { score: max_score, student_message: 'success' }

Author: ```suggestion
          { score: maximum_score, student_message: 'success' }
```

        end
```

##### Code

The code block must be valid Ruby code. It must have a method called `check`, and may optionally contain helper methods as well.

> The method `check` will be called with two arguments:
> - A map from `[RESOURCE].[SERVICE]` to a handle to that resource's service. It will only contain handles to services specified in the step's `services` array.
> - The maximum score for the step.

Reviewer 2: ```suggestion
The method `check` will be called with three keyword arguments:

- `handles`: A map from `[RESOURCE].[SERVICE]` to a handle to that resource's service. It will only contain handles to services specified in the step's `services` array.
- `resources`: A map from `[RESOURCE]` (ids) to maps of each ....
- `maximum_score`: The maximum score for the step.
```

The method `check` should return a single hash with:
- `:score`: the number of points the student earned.

> - `:student_message`: a key from the step's `student_messages` array, which will be presented to the student in the appropriate locale.

Reviewer 2: - [ ] TODO(joshgan): Student message should be passed through if it's not found in the locale hash.
- [ ] TODO(joshgan): Do we need to both `message` and `student_message`. It's unclear how `message` is currently used.
